### ISSUE TITLE
Promise based support for Template.prototype.evaluate

### ISSUE THREAD

**Issue by umaar** [nodeca/mincer#233]
I was hoping to make a mincer plugin for [Rollup](https://rollupjs.org/), however their API is promise based. Could you support this in `Template.prototype.evaluate`?

Right now, if I specify an Promise based function for `evaluate`, [`context.evaluate(this.pathname)`](https://github.com/nodeca/mincer/blob/a973f84e39ffb48fd725032494e0c0d14e933b55/lib/mincer/assets/processed.js#L145) won't wait for it.

Also see this Stackoverflow question: http://stackoverflow.com/questions/21424867/run-an-asynchronous-stream-based-function-synchronously-in-node-js

**Comment by umaar** [comment]
I should add: even if it's not promise based, at least a callback would be nice. I don't see how any async code can be used in `evaluate`

**Comment by puzrin** [comment]
No way now. And since this package is not actively maintained, i don't think this feature will appear.

**Comment by umaar** [comment]
That's a shame, but thanks for the input.

**Comment by puzrin** [comment]
See https://github.com/nodeca/mincer/issues/150 for explanations. Initial concept (sprockets port) is not good for component-based approach, because too file-centric. Probably, this can be reworked as independent package, but i have no resource for that. If anyone wish to continue, i can add permissions and make full permissions transfer later (if everything goes well).

**Comment by umaar** [comment]
Ok good to know. I'm interested in this file: [bundle.js](https://github.com/nodeca/nodeca.core/blob/master/lib/system/init/bundle.js) file, it looks like you're using it as an alternative to mincer. 

Do you find it works better? What are your thoughts on some of the other build tools out there like Gulp/webpack etc. which can also do all of those things via plugins - and also allow programmatic usage.

**Comment by puzrin** [comment]
> Do you find it works better?

It's better for __my__ needs, because i don't have to think about all possible use cases, engines, extensions and documentation.

> What are your thoughts on some of the other build tools out there like Gulp/webpack etc. which can also do all of those things via plugins - and also allow programmatic usage.

IMHO, that works well only until you have simple requirements. When you need something more, you will spend half of time to fight with bad api and push your needs to upstream :).

I'm sure, this can be generalized somehow, but this requires significant time and experience. At this time i don't know good package that can help to solve this problem. When it appears, i'll be happy to use it.

**Comment by umaar** [comment]
Fair enough, I admire the effort you've put into, for example, [plugins](https://github.com/nodeca/nodeca.core/tree/master/lib/bundler/plugins), it's interesting to see. Thanks for explaining!

**Comment by puzrin** [comment]
If you decide to dig nodeca bundler or create your own, pay attention to [macros](https://github.com/nodeca/nodeca.core/blob/master/lib/bundler/plugins/macros.js) implementation and related things. The most painful problem is that you may wish to have sync helpers in html templates, but async imlementations of plugins. If helper's dependencies (asset's url for example) are not ready, you can't silently build those after helper invoked.

Solution is to throw error "unfulfiled dependencies", build missed ones and rerun build of resource that caused error. At least, that worked with my use cases. That's the main logical collision you should care if you decide to create complicated assets bundler.
